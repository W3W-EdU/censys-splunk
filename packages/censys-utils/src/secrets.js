import {
    config,
    createRESTURL,
    defaultFetchInit,
    fetch,
    getEntries,
    makeJsonUrl,
    raiseOnError,
} from './api';

const DEFAULT_SECRET_NAME = 'censys_secrets';
const DEFAULT_REALM = 'censys_setup';
const ASM_API_KEY_REGEX = /^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$/;
const SEARCH_API_ID_REGEX = ASM_API_KEY_REGEX;
const SEARCH_API_SECRET_REGEX = /^[a-z0-9]{32}$/i;

const mapEntryToSecretEntry = (entry) => {
    const { name, content } = entry;
    if (content) {
        const { username, realm } = content;
        let clearPassword = content.clear_password;
        try {
            clearPassword = JSON.parse(clearPassword);
        } catch (error) {
            // do nothing
        }
        return {
            title: name,
            username,
            clearPassword,
            realm,
        };
    }
    return null;
};

const mapEntriesToSecretEntries = (entries) => {
    return entries.map(mapEntryToSecretEntry).filter((entry) => entry !== null);
};

const getSecretEntries = (app = config.app, signal = null) => {
    const url = createRESTURL('storage/passwords', {
        app,
        sharing: 'app',
    });

    return fetch(makeJsonUrl(url), { ...defaultFetchInit, signal })
        .then(raiseOnError)
        .then((response) => response.json())
        .then(getEntries)
        .then(mapEntriesToSecretEntries);
};

const getSecretEntry = (secretName, app = config.app, signal = null) => {
    const defaultErrorMsg = 'Unable to get secret entry';

    return getSecretEntries(signal, app).then((secretEntries) => {
        const secretEntry = secretEntries.find((entry) => entry.username === secretName);
        if (!secretEntry) {
            throw new Error(defaultErrorMsg);
        }
        return secretEntry;
    });
};

const createSecretEntry = (secretName, secrets, app = config.app, signal = null) => {
    const url = createRESTURL('storage/passwords', {
        app,
        sharing: 'app',
    });

    // Create form data to send
    const data = new URLSearchParams();
    data.append('name', secretName);
    data.append('password', JSON.stringify(secrets));
    data.append('realm', DEFAULT_REALM);

    // Make a POST request to set the secret entries
    return fetch(makeJsonUrl(url), {
        ...defaultFetchInit,
        signal,
        method: 'POST',
        body: data,
    })
        .then(raiseOnError)
        .then((response) => response.json())
        .then(getEntries)
        .then(mapEntriesToSecretEntries)
        .then((secretEntries) => {
            if (secretEntries.length === 0) {
                throw new Error('Unable to set secret entries');
            }
            return secretEntries[0];
        });
};

const updateSecretEntry = (secretName, secrets, app = config.app, signal = null) => {
    const url = createRESTURL(`storage/passwords/${DEFAULT_REALM}:${secretName}:`, {
        app,
        sharing: 'app',
    });

    // Create form data to send
    const data = new URLSearchParams();
    data.append('password', JSON.stringify(secrets));

    // Make a POST request to update the secret entries
    return fetch(makeJsonUrl(url), {
        ...defaultFetchInit,
        signal,
        method: 'POST',
        body: data,
    })
        .then(raiseOnError)
        .then((response) => response.json())
        .then(getEntries)
        .then(mapEntriesToSecretEntries)
        .then((secretEntries) => {
            if (secretEntries.length === 0) {
                throw new Error('Unable to update secret entries');
            }
            return secretEntries[0];
        });
};

export default {
    DEFAULT_SECRET_NAME,
    DEFAULT_REALM,
    ASM_API_KEY_REGEX,
    SEARCH_API_ID_REGEX,
    SEARCH_API_SECRET_REGEX,
    getSecretEntries,
    getSecretEntry,
    createSecretEntry,
    updateSecretEntry,
};
