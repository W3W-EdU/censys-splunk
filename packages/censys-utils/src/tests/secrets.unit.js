/* eslint-env jest */
import { assert } from 'chai';

import * as url from '@splunk/splunk-utils/url';

import secrets from '../secrets';
import { ERROR, PASSWORD, PASSWORDS, UPDATE_ERROR, makeUrl } from './utils';

const ENTRY_TITLE = `${secrets.DEFAULT_REALM}:${secrets.DEFAULT_SECRET_NAME}:`;
const TEST_SECRETS = {
    censys_asm_api_key: 'test',
    censys_search_app_id: 'test',
    censys_search_app_secret: 'test',
};

jest.spyOn(url, 'createRESTURL').mockImplementation(makeUrl);

const assertEntryIsValid = (
    entry,
    expectedTitle = ENTRY_TITLE,
    expectedUsername = secrets.DEFAULT_SECRET_NAME,
    expectPassword = true
) => {
    assert.isObject(entry);
    assert.equal(entry.title, expectedTitle);
    assert.equal(entry.username, expectedUsername);
    if (expectPassword) {
        assert.isObject(entry.clearPassword);
        assert.isString(entry.clearPassword.censys_asm_api_key);
        assert.isString(entry.clearPassword.censys_search_app_id);
        assert.isString(entry.clearPassword.censys_search_app_secret);
    } else {
        assert.isUndefined(entry.clearPassword);
    }
};

describe('CensysSetupUtils', () => {
    beforeEach(() => {
        fetch.resetMocks();
    });
    it('gets secret entries', async () => {
        fetch.mockResponseOnce(PASSWORDS, {
            status: 200,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl('storage/passwords'),
        });
        const secretEntries = await secrets.getSecretEntries();
        assert.equal(secretEntries.length, 1);
        const secretEntry = secretEntries[0];
        assertEntryIsValid(secretEntry);
    });
    it('error getting secret entries', async () => {
        fetch.mockResponseOnce(ERROR, {
            status: 500,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl('storage/passwords'),
        });
        await expect(secrets.getSecretEntries()).rejects.toThrowErrorMatchingSnapshot();
    });
    it('gets secret entry', async () => {
        fetch.mockResponseOnce(PASSWORDS, {
            status: 200,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl('storage/passwords'),
        });
        const secretEntry = await secrets.getSecretEntry(secrets.DEFAULT_SECRET_NAME);
        assertEntryIsValid(secretEntry);
    });
    it('error getting secret entry', async () => {
        fetch.mockResponseOnce(ERROR, {
            status: 500,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl('storage/passwords'),
        });
        await expect(
            secrets.getSecretEntry(secrets.DEFAULT_SECRET_NAME)
        ).rejects.toThrowErrorMatchingSnapshot();
    });
    it('creates secret entry', async () => {
        fetch.mockResponseOnce(PASSWORD, {
            status: 200,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl(`storage/passwords/${ENTRY_TITLE}`),
        });
        const secretEntry = await secrets.createSecretEntry(
            secrets.DEFAULT_SECRET_NAME,
            TEST_SECRETS
        );
        assertEntryIsValid(secretEntry, ENTRY_TITLE, secrets.DEFAULT_SECRET_NAME, false);
    });
    it('error creating secret entry', async () => {
        fetch.mockResponseOnce(ERROR, {
            status: 404,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl(`storage/passwords/${ENTRY_TITLE}`),
        });
        await expect(
            secrets.createSecretEntry(secrets.DEFAULT_SECRET_NAME, TEST_SECRETS)
        ).rejects.toThrowErrorMatchingSnapshot();
    });
    it('updates secret entry', async () => {
        fetch.mockResponseOnce(PASSWORD, {
            status: 200,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl(`storage/passwords/${ENTRY_TITLE}`),
        });
        const secretEntry = await secrets.updateSecretEntry(
            secrets.DEFAULT_SECRET_NAME,
            TEST_SECRETS
        );
        assertEntryIsValid(secretEntry, ENTRY_TITLE, secrets.DEFAULT_SECRET_NAME, false);
    });
    it('error updating secret entry', async () => {
        fetch.mockResponseOnce(UPDATE_ERROR, {
            status: 404,
            headers: { 'Content-Type': 'application/json' },
            url: makeUrl(`storage/passwords/${ENTRY_TITLE}`),
        });
        await expect(
            secrets.updateSecretEntry(secrets.DEFAULT_SECRET_NAME, TEST_SECRETS)
        ).rejects.toThrowErrorMatchingSnapshot();
    });
});
