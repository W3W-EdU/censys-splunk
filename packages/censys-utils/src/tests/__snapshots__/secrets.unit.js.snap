// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CensysSetupUtils error creating secret entry 1`] = `"404 Not Found: https://splunk.dev/service/storage/passwords/censys_setup:censys_secrets:"`;

exports[`CensysSetupUtils error getting secret entries 1`] = `"500 Internal Server Error: https://splunk.dev/service/storage/passwords"`;

exports[`CensysSetupUtils error getting secret entry 1`] = `"500 Internal Server Error: https://splunk.dev/service/storage/passwords"`;

exports[`CensysSetupUtils error updating secret entry 1`] = `"404 Not Found: https://splunk.dev/service/storage/passwords/censys_setup:censys_secrets:"`;
